### ISSUE TITLE
Improve docs + error messages for companion objects of ZIO type aliases

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#744]
ZIO has a range of type aliases, including:

 - `IO`
 - `UIO`
 - `Task`
 - `TaskR`

These type aliases have companion objects, which are named identically.

Currently, type bounds are used to ensure that, for example, all methods in the `UIO` companion object may NOT create fallible effects.

However, the error message that one gets when one tries to create a fallible effect is quite terrible:

```scala
inferred type arguments [String] do not conform to method fail's type parameter bounds [E <: scalaz.zio.UIO.UpperE]
```

This error message is non-sensical and doesn't guide users to successful resolution of the problem. Indeed, despite knowing how ZIO works and having written a lot of the code, I am momentarily confused by the error that is generated when I accidentally call `IO.bracket` and use an environmental effect, which would call for `ZIO.bracket`.

To solve this problem, there are two options that I can see:

1. Use implicits to demand compliance with the requirements of the type alias. This is the approach utilized in `ZSchedule`, for example. The advantage to implicits is that the `@implicitNotFound` annotation can be used, which can be customized on a per-companion object basis, and which can provide a _quite detailed and helpful_ message to the user.
2. Abandon the current approach and generate companion objects with replicated methods and hand-coded type signatures, which delegate to the more general effect methods in the ZIO companion object. This would eliminate all implicits and all type bounds, at the cost of introducing more boilerplate-y code, and increasing maintenance a bit (making sure to keep them in sync).

Overall I am not sure which approach is better, but I do feel like (2) will make it easier for new contributors and solve associated duplication problems, because figuring out the type bounds or the implicits can be quite confusing at first glance.

However, if (1) is chosen, then we need to perform an additional step: using the `@usecase` annotation in Scaladoc to hide the implicits, because they are not required to understand the methods and they are provided already by the companion objects. Currently the method signatures in the generated Scaladoc are quite confusing because of all the type bound noise.

I'm leaning toward (2), making new files for `UIO`, `Task`, `TaskR`, and `IO`, which merely delegate to the `ZIO` methods, because it results in the clearest organization without Scaladoc hacks and is immediately obvious to new contributors; albeit, it will require care during pull requests to ensure if a method is added to `ZIO`, it is added to the applicable companion objects.

**Comment by heraklos** [comment]
Hello @jdegoes 
I would like to take this issue.

**Comment by TheMayoras** [comment]
Hello @jdegoes and @heraklos, I would also like to help with this issue.

**Comment by heraklos** [comment]
@TheMayoras 
Hi, thanks for your message. Please take this issue 😄 
It seems I won't be able to give enough time to work on this issue this month.

**Comment by TheMayoras** [comment]
Hello @jdegoes,
I'm not sure if this is the right place to ask, but I am hoping to make this issue my first contribution to an open source project.  It seems doable, but I was hoping for a bit of help getting started.  I've read through most of the documentation, but I think some extra information would help me get started.

Thank You 😊

**Comment by jdegoes** [comment]
@TheMayoras Absolutely! Thanks for your interest in this one.

Would you like me to write more explanation here, or have a short video session? Or both? 😄

**Comment by TheMayoras** [comment]
@jdegoes An explanation should hopefully be enough.  If I understand correctly, the goal for (2) is to create a new file with "improved" IO, UIO, Task, and TaskR objects that have more specific type signatures that shuffle the methods to ZIO.  Is that on the right track?  I'm sure I am missing a lot of details here.

Thank you for the guidance. I have been nervous to about contributing to anything open-source for a while.

**Comment by jdegoes** [comment]
@TheMayoras 

The ticket can be structured in the following phases.

### Phase 1: Introduce UIO.scala, TaskR.scala, Task.scala, and IO.scala

Initially, these will be empty files, but soon they will contain all the "companion objects" for the type aliases `UIO`, `TaskR`, `Task`, and `IO`.

### Phase 2: Move the `UIO`, `TaskR`, `Task`, and `IO` objects into their respective files.

A simple migration of objects out of `ZIO.scala` into the newly-created files.

### Phase 3: Copy/paste all the methods that each companion object inherits into the companion objects directly, and completely remove the inheritance.

Currently, each object inherits from different traits. You have to look at all those traits and copy / paste the methods from them into the companion objects. For example, copy/paste `succeed` from `ZIOFunctions` trait into the companion objects: `UIO`, `Task`, `TaskR`, and `IO`. When you are done with this step, the companion objects won't inherit from anything anymore.

### Phase 4: Have all implementations of all methods inside the companion objects _delegate_ to the methods in `ZIO.*`. 

Technically you could do this at the same time as (3), and maybe it would be simpler that way. There's no need to have more than one implementation for anything in here, so they can all delegate to the "most powerful" variant in the ZIO companion object.

### Phase 5: Remove all type bounds / `UpperE` / `LowerE` from `ZIO.scala` and the other effect companion object files.

All this type boundary machinery can go!

### Phase 6: Specialize the `R` and `E` type parameters inside UIO.scala, TaskR.scala, Task.scala, and IO.scala.

For example, in `Task.scala`, all methods will be specialized with `E = Throwable` and `R = Any`. While in `TaskR.scala`, all methods will be specialized with `E = Throwable`, but `R` will remain polymorphic. In `IO.scala`, all methods will be specialized with `R = Any`. In `UIO.scala`, all methods will be specialized with `E = Nothing`, and `R = Any`.

### Phase 7: Add Scaladoc links from companion object methods to ZIO methods.

For example, in `Task.succeed`, there can be a link to `[[ZIO.succeed]]`. To avoid duplicating Scaladoc across all the methods.

### Phase 8: Inline all ZIO methods in `ZIO.scala` into the ZIO companion object, and delete the existing traits (like `ZIOFunctions`).

After this step is completed, there will only exist ZIO companion object and the other companion objects. No more "helper" traits filled with effect methods.

### Phase 9: Sort all methods in companion objects alphabetically.

This will help us keep things in sync moving forward and is good for organization.

**Comment by jdegoes** [comment]
@TheMayoras Let me know if you need anything more, and I'm really excited you chose ZIO to be your first OSS contribution! (Hopefully it will be a great experience that results in more contributions. 😄 )

**Comment by TheMayoras** [comment]
@jdegoes I just want to clarify something about **Phase 6** before I go any further.  For Task.scala for example, do you want all _E_ to be `Throwable`, or all _E_ to be `E<:Throwable`.  

I was also wondering what you thought about something like a **_Guide to the Code for Beginners_**.  Explaining the reasoning behind decisions, how the code fits together, etc.  I know it would help me.

**Comment by jdegoes** [comment]
@TheMayoras For `Task`, all `E` must be `Throwable`. For non-`Throwable`, a user must use `IO`, for example.

That's a great idea! Unfortunately I won't have time to work on that until the late summer...

**Comment by TheMayoras** [comment]
@jdegoes I would love to help write it.  I would obviously need a lot of questions answered 😂

**Comment by TheMayoras** [comment]
@jdegoes I am working on **Phase 6** on `Task.scala` right now, and I am seeing some functions that take multiple `R`'s.  For example, `raceAll` takes an `R` which I changed to `Throwable`, but it also take an `Iterable[ZIO[R1>: LowerR >: R, E, A]]`.  Should I convert `R1` to `Throwable` or change it to `R1>:`  `Throwable`?

**Comment by jdegoes** [comment]
I'm good to answer questions!

For `Task`, all `R` must be converted to `Any`, so there will not be any more `R` type parameters in `Task.scala`.

**Comment by TheMayoras** [comment]
@jdegoes  Sorry for all the questions, but what type of `E` should I use for `IO`?

**Comment by jdegoes** [comment]
`E` will have no constraints and be fully polymorphic for `IO`. Only `R` will be replaced by `Any` in `IO`.

For `ZIO`, everything will be fully polymorphic for `R` and `E`.

**Comment by jdegoes** [comment]
No worries on questions, that's what I'm here for. 😄

**Comment by TheMayoras** [comment]
@jdegoes I have just finished with **Phase 6**, and when I try to compile, I get an error saying:
`D:\Software\Eclipse\Workspace\Scala\scalaz-zio_issue744\scalaz-zio\core\shared\src\main\scala\scalaz\zio\Promise.scala:56:28:` 
`wrong number of type parameters for method effectAsyncInterrupt:`
`[E, A](register: (scalaz.zio.ZIO[scalaz.zio.IO.IoR,E,A] => Unit) => Either[scalaz.zio.Canceler,scalaz.zio.ZIO[scalaz.zio.IO.IoR,E,A]])scalaz.zio.ZIO[scalaz.zio.IO.IoR,E,A]
[error]     IO.effectAsyncInterrupt[Any, E, A](k => {`.  
It seems to me that I could either remove the initial `Any` in `Promise` or add an extra `R` in `IO`,.  Removing `Any` from `Promise` is a better solution.  How would you like me to go from here?

**Comment by jdegoes** [comment]
@TheMayoras Can you submit a pull request so I can take a more detailed look at the code?

I think since `IO` will not take `R` as a type parameter, you can just delete that `Any` in the call to `effectAsyncInterrupt`.

**Comment by TheMayoras** [comment]
@jdegoes Just created a pull request.

**Comment by ghostdogpr** [comment]
Closed by #805
